# Question 1


Representing Data as a 4 x 3 Matrix

scores = rows and subjects = column

$ scores=\begin{bmatrix}80&70&90\cr60&85&75\cr95&88&92\cr70&60&65\end{bmatrix}$

In [8]:
import numpy as np

# create 4 x 3 matrix 
scores = np.array([[80, 70, 90],[60, 85, 75], 
                  [95, 88, 92], [70, 60, 65]])
print("Student scores in a matrix:\n", scores)



Student scores in a matrix:
 [[80 70 90]
 [60 85 75]
 [95 88 92]
 [70 60 65]]


Total Scores per Student(vector addition)

Total_score_per_student $=\begin{bmatrix}80 + 70 +90&\cr60 +85 +75\cr95 +88 + 92\cr70 +60 +65\end{bmatrix}$ =
$\begin{bmatrix}240&\cr220\cr275\cr195\end{bmatrix}$

In [2]:
print("Total score for each (student_1, student_2, student_3, student_4) :", np.sum(scores, axis=1))  
print("\n")
print("Student_1_score:", np.sum(scores[:1]))
print("\n")
print("Student_2_score:", np.sum(scores[1:2]))
print("\n")
print("Student_3_score:", np.sum(scores[2:3]))
print("\n")
print("Student_4_score:", np.sum(scores[3:4]))

Total score for each (student_1, student_2, student_3, student_4) : [240 220 275 195]


Student_1_score: 240


Student_2_score: 220


Student_3_score: 275


Student_4_score: 195


Applying weight (scalar multiplication)

Weights  = 2 x math

Weighted_value  $ = \begin{bmatrix}2 * 80&70&90\cr2 * 60&85&75\cr2 * 95&88&92\cr2 * 70&60&65\end{bmatrix}$ 
$ = \begin{bmatrix}160&70&90\cr120&85&75\cr190&88&92\cr140&60&65\end{bmatrix}$

In [15]:
weight = np.array([[2, 0, 0],[0, 1, 0],[0, 0, 1]])
print("Weight using diagonal matrix:\n", weight)
print("\n")
subject_weight = scores @ weight 
print("Student scores:\n", subject_weight)

# scores = scores.copy()
# scalar_math = 2

# scores[:, 0] = scalar_math * scores[:, 0]
# print("Student scores:\n", scores)

Weight using diagonal matrix:
 [[2 0 0]
 [0 1 0]
 [0 0 1]]


Student scores:
 [[160  70  90]
 [120  85  75]
 [190  88  92]
 [140  60  65]]


Average Score per subject (Matrix operation)

Average_score_ per_Subject $= \begin{bmatrix}(80 + 60 + 95 + 70)\div 4 &\cr(70 +85 + 88 + 60)\div 4\cr(90 + 75 + 92 + 65)\div 4\end{bmatrix}$ =
$\begin{bmatrix}76.25&\cr75.75\cr80.5\end{bmatrix}$

In [16]:
print("Average score for each (Math, English, Science) :", np.mean(scores, axis=0))  

Average score for each (Math, English, Science) : [76.25 75.75 80.5 ]


Final Grade using matrix multiplication

Weight = $\begin{bmatrix}0.5&\cr0.3\cr0.2\end{bmatrix}$

Final_grades = $\begin{bmatrix}80&70&90\cr60&85&75\cr95&88&92\cr70&60&65\end{bmatrix}$ X $\begin{bmatrix}0.5&\cr0.3\cr0.2\end{bmatrix}$ = $\begin{bmatrix}79.0&\cr70.5\cr92.3\cr66.0\end{bmatrix}$ 

In [18]:

weight_vector = np.array([[0.5], [0.3], [0.2]])
final_grades = scores @ weight_vector
#final_grade = np.dot(scores, weight_vector)
print("The final_grades:", final_grades)

The final_grades: [[79. ]
 [70.5]
 [92.3]
 [66. ]]


Comparing Students  (Vector Subtraction)

Student_1 Score vector = $\begin{bmatrix}80&70&90\end{bmatrix}$

Student_2 Score vector = $\begin{bmatrix}60&85&75\end{bmatrix}$

Student Comparison (1 & 2) = $\begin{bmatrix}80&70&90\end{bmatrix}$ - $\begin{bmatrix}60&75&85\end{bmatrix}$

Student Comparison (1 & 2) = $\begin{bmatrix}20&-15&15\end{bmatrix}$

In Math - student 1 performs better than student 2 by scoring 20 marks higher 

In English - student 2 performs better than student 1 by scoring 15 marks higher

In Science - student 1 performs better than student 2 by scoring 15 marks higher 

In conclusion: Student 1 performs better in two subject compare to Student 2 that perform better in one subject

In [21]:
student_1_score = scores[0:1]
student_2_score = scores[1:2]

print("Student_1 Score vector:",student_1_score)
print("\n")
print("Student_2 Score vector:",student_2_score)

compare_scores = student_1_score - student_2_score
print("\n")
print("comparing student (1 & 2) vector score", compare_scores)

Student_1 Score vector: [[80 70 90]]


Student_2 Score vector: [[60 85 75]]


comparing student (1 & 2) vector score [[ 20 -15  15]]


# Question 2

Representing Data as a 5 x 4 Matrix

scores = rows and subjects = column

new_scores= $\begin{bmatrix}82&71&77&55\cr50&45&79&82\cr59&68&72&53\cr30&70&55&64\cr72&83&65&87\end{bmatrix}$

In [36]:
new_scores = np.array([[82, 71, 77, 55],[50, 45, 79, 82], 
                  [59, 68, 72, 53], [30, 70, 55, 64], [72, 83, 65, 87]])
print("Student scores in a matrix:", new_scores)

Student scores in a matrix: [[82 71 77 55]
 [50 45 79 82]
 [59 68 72 53]
 [30 70 55 64]
 [72 83 65 87]]


In [100]:
new_scores.shape

(5, 4)

Total Scores per Student(vector addition)

Total_score_per_student $=\begin{bmatrix}82 + 71 + 77 + 55&\cr50 + 45 + 79 + 82\cr59 + 68 + 72 + 53\cr30 +70 +55 + 64\cr72 + 83 +65 + 87\end{bmatrix}$ =
$\begin{bmatrix}285&\cr256\cr252\cr219\cr307\end{bmatrix}$

In [37]:
print("Total score for each (student_1, student_2, student_3, student_4, student_5) :", np.sum(new_scores, axis=1)) 

print("\n")
print("Student_1_score:", np.sum(new_scores[:1]))
print("\n")
print("Student_2_score:", np.sum(new_scores[1:2]))
print("\n")
print("Student_3_score:", np.sum(new_scores[2:3]))
print("\n")
print("Student_4_score:", np.sum(new_scores[3:4]))
print("\n")
print("Student_5_score:", np.sum(new_scores[4:5]))

Total score for each (student_1, student_2, student_3, student_4, student_5) : [285 256 252 219 307]


Student_1_score: 285


Student_2_score: 256


Student_3_score: 252


Student_4_score: 219


Student_5_score: 307


Applying weight (scalar multiplication)

Weights  = 3 x Science

Weighted_value  = $\begin{bmatrix}82&71&77 * 3&55\cr50&45&79 * 3&82\cr59&68&72 * 3&53\cr30&70&55 * 3&64\cr72&83&65 * 3&87\end{bmatrix}$ = $\begin{bmatrix}82&71&231&55\cr50&45&237&82\cr59&68&216&53\cr30&70&165&64\cr72&83&195&87\end{bmatrix}$ 

In [38]:
new_weight = np.array([[1, 0, 0, 0],[0, 1, 0, 0],
                   [0, 0, 3, 0], [0, 0, 0, 1]])
print("Weight using diagonal matrix:\n", new_weight)
print("\n")
new_subject_weight = new_scores @ new_weight
print("Student scores:\n", new_subject_weight)




# new_scores = new_scores.copy()
# scalar_science = 3

# new_scores[:, 2] = scalar_science * new_scores[:, 2]
# print("new scores:\n", new_scores)
# print("\n")

Weight using diagonal matrix:
 [[1 0 0 0]
 [0 1 0 0]
 [0 0 3 0]
 [0 0 0 1]]


Student scores:
 [[ 82  71 231  55]
 [ 50  45 237  82]
 [ 59  68 216  53]
 [ 30  70 165  64]
 [ 72  83 195  87]]


Average Score per subject (Matrix operation)

Average_score_ per_Subject $= \begin{bmatrix}(82 + 50 + 59 + 30 + 72)\div 5 &\cr(71 + 45 + 68 + 70 + 83)\div 5\cr(77 + 79 + 72 + 55 + 65)\div 5\cr(55 + 82 + 53 + 64 + 87)\div 5\end{bmatrix}$ =
$\begin{bmatrix}58.6&\cr67.4\cr69.6\cr68.2\end{bmatrix}$

In [39]:
print("Average score for each (Math, English, Science, chemistry) :", np.mean(new_scores, axis=0)) 

Average score for each (Math, English, Science, chemistry) : [58.6 67.4 69.6 68.2]


In [128]:
new_scores

array([[82, 71, 77, 55],
       [50, 45, 79, 82],
       [59, 68, 72, 53],
       [30, 70, 55, 64],
       [72, 83, 65, 87]])

Final Grade using matrix multiplication

Weight = $\begin{bmatrix}0.5&\cr0.3\cr0.2\end{bmatrix}$

Final_grades = $\begin{bmatrix}82&71&77&55\cr50&45&79&82\cr59&68&72&53\cr30&70&55&64\cr72&83&65&87\end{bmatrix}$ X $\begin{bmatrix}0.4&\cr0.2\cr0.3\cr0.1\end{bmatrix}$ = $\begin{bmatrix}75.6&\cr60.9\cr64.1\cr48.9\cr73.6\end{bmatrix}$ 

In [42]:
weight_vector_2 = np.array([[0.4], [0.2], [0.3], [0.1]])
print(weight_vector_2.shape)
print("\n")
final_grades_2 = new_scores @ weight_vector_2
#final_grade = np.dot(scores, weight_vector)
print("The final_grades:", final_grades_2)

(4, 1)


The final_grades: [[75.6]
 [60.9]
 [64.1]
 [48.9]
 [73.6]]


Comparing Students  (Vector Subtraction)

Student_3 Score vector = $\begin{bmatrix}59&68&72&53\end{bmatrix}$

Student_4 Score vector = $\begin{bmatrix}30&70&55&64\end{bmatrix}$

comparing student (3 & 4) vector score = $\begin{bmatrix}29&-2&17&-11\end{bmatrix}$

In math - student 3 performs better

In english - student 4 performs better

In Science - student 3 performs better

In chemistry - student 4 performs better

In [129]:
student_3_score = new_scores[2:3]
student_4_score = new_scores[3:4]

print("Student_3 Score vector:",student_3_score)
print("\n")
print("Student_4 Score vector:",student_4_score)

compare_scores_2 = student_3_score - student_4_score
print("\n")
print("comparing student (3 & 4) vector score", compare_scores_2)

Student_3 Score vector: [[59 68 72 53]]


Student_4 Score vector: [[30 70 55 64]]


comparing student (3 & 4) vector score [[ 29  -2  17 -11]]
